src/chords_generation/daria.py

```diff
@@ -0,0 +1,126 @@
+from typing import List, Tuple
+
+import music
+import utils.constants
+from . import daria_params as params
+
+# https://www.fim.uni-passau.de/fileadmin/dokumente/fakultaeten/fim/lehrstuhl/sauer/geyer/BA_MA_Arbeiten/BA-HausnerChristoph-201409.pdf
+'''
+Split notes into timeline
+Split timeline into windows (bars)
+    Extend some bars to prevent changing chord in the middle of sound
+For each bar get best chord using scoring lists
+'''
+
+
+def tonation_chords(tonation: music.Tonation):
+    n = tonation.note
+    if tonation.kind == "major":
+        chords = [
+            music.Chord(note=n, kind="major"),
+            music.Chord(note=n+2, kind="minor"),
+            music.Chord(note=n+4, kind="major"),
+            music.Chord(note=n+5, kind="major"),
+            music.Chord(note=n+7, kind="major7"),
+            music.Chord(note=n+9, kind="minor"),
+            music.Chord(note=n+11, kind="diminished")]
+    else:
+        chords = [
+            music.Chord(note=n, kind="minor"),
+            music.Chord(note=n+2, kind="diminished"),
+            music.Chord(note=n+4, kind="major"),
+            music.Chord(note=n+5, kind="minor"),
+            music.Chord(note=n+7, kind="major7"),
+            music.Chord(note=n+9, kind="major"),
+            music.Chord(note=n+11, kind="diminished")]
+    for i in range(len(chords)):
+        chords[i].level = i+1
+    return chords
+
+
+def points_for_sound(chord: music.Chord, sound: music.Sound):
+    if sound.note is None:
+        return 0
+    diff = (sound.note+12 - chord.note) % 12
+    return params.POINTS[chord.kind][diff]
+
+
+def point_coef(chord: music.Chord, first_sound: music.Sound, last_chord_level: int):
+    if chord.note == first_sound.note:
+        first_sound_coef = params.FIRST_SOUND
+    else:
+        first_sound_coef = 1
+    return params.COEFS[chord.kind] * \
+        params.NEXT_CHORD[last_chord_level].get(chord.level, 1) * \
+        first_sound_coef
+
+
+def get_chords_daria(sounds: List[music.Sound],
+                     tonation: music.Tonation,
+                     meter: Tuple[int, int]):
+    # set meter
+    meter = (
+        meter[0],
+        utils.constants.RHYTHMIC_VALUE_TO_TIME[str(meter[1])]
+    )
+    if meter[0] == 2:
+        meter = (
+            4,
+            meter[1]/2
+        )
+    if meter[0] != 4:
+        raise Exception("Not supported meter in chord duration")
+
+    # allow changing chords every half bar
+    half_bar_len = 2*int(meter[1]/0.125)
+
+    # create timeline of sounds
+    t_s = []
+    for s in sounds:
+        for _ in range(int(s.rhythmic_value_time*8)):
```

Comment:
Oh, ok, so this is a way to ensure that all times are whole numbers.
This is tragic. RHYTHMIC_VALUE_TO_TIME is completely arbitrary (and whatever the result is, it is NOT time, rather some sort of relation to the whole note), and consequences of this idiocy strech across the whole project, generating this type of magic numbers just everywhere. It is one of the reasons our code is so hard to understand.

I propose to store `rhytmic_value` as integers which are effectively "number of smallest notes". For some reason we decided to use "4." notation that will be useful only when passing the data to frontend, and we struggle with it throughout the whole project. I think whis change should be done just after this PR. What do you think (is this change worth our time in your opinion)?

Reply:
We can do this